pyteal/ast/router.py

```diff
@@ -0,0 +1,541 @@
+                "generate condition based on CallConfigs should be based on CallConfig"
+            )
+        config_oc_pairs: list[tuple[CallConfig, EnumInt]] = [
+            (self.no_op, OnComplete.NoOp),
+            (self.opt_in, OnComplete.OptIn),
+            (self.close_out, OnComplete.CloseOut),
+            (self.clear_state, OnComplete.ClearState),
+            (self.update_application, OnComplete.UpdateApplication),
+            (self.delete_application, OnComplete.DeleteApplication),
+        ]
+        return [
+            oc
+            for oc_config, oc in config_oc_pairs
+            if (oc_config & call_config) != CallConfig.NEVER
+        ]
+
+    def partition_oc_by_clear_state(self, cc: CallConfig) -> tuple[bool, list[EnumInt]]:
+        not_clear_states, clear_states = more_itertools.partition(
+            lambda x: str(x) == str(OnComplete.ClearState),
+            self._oc_under_call_config(cc),
+        )
+        return len(list(clear_states)) > 0, list(not_clear_states)
+
+
+@dataclass(frozen=True)
+class OnCompleteAction:
+    """
+    OnComplete Action, registers bare calls to one single OnCompletion case.
+    """
+
+    on_create: Optional[Expr | SubroutineFnWrapper | ABIReturnSubroutine] = field(
+        kw_only=True, default=None
+    )
+    on_call: Optional[Expr | SubroutineFnWrapper | ABIReturnSubroutine] = field(
+        kw_only=True, default=None
+    )
+
+    @staticmethod
+    def never() -> "OnCompleteAction":
+        return OnCompleteAction()
+
+    @staticmethod
+    def create_only(
+        f: Expr | SubroutineFnWrapper | ABIReturnSubroutine,
+    ) -> "OnCompleteAction":
+        return OnCompleteAction(on_create=f)
+
+    @staticmethod
+    def call_only(
+        f: Expr | SubroutineFnWrapper | ABIReturnSubroutine,
+    ) -> "OnCompleteAction":
+        return OnCompleteAction(on_call=f)
+
+    @staticmethod
+    def always(
+        f: Expr | SubroutineFnWrapper | ABIReturnSubroutine,
+    ) -> "OnCompleteAction":
+        return OnCompleteAction(on_create=f, on_call=f)
+
+    def is_empty(self) -> bool:
+        return not (self.on_call or self.on_create)
+
+
+OnCompleteAction.__module__ = "pyteal"
+
+
+@dataclass(frozen=True)
+class OnCompleteActions:
+    """
+    OnCompletion Actions keep track of bare-call registrations to all OnCompletion cases.
+    """
+
+    close_out: OnCompleteAction = field(kw_only=True, default=OnCompleteAction.never())
+    clear_state: OnCompleteAction = field(
+        kw_only=True, default=OnCompleteAction.never()
+    )
+    delete_application: OnCompleteAction = field(
+        kw_only=True, default=OnCompleteAction.never()
+    )
+    no_op: OnCompleteAction = field(kw_only=True, default=OnCompleteAction.never())
+    opt_in: OnCompleteAction = field(kw_only=True, default=OnCompleteAction.never())
+    update_application: OnCompleteAction = field(
+        kw_only=True, default=OnCompleteAction.never()
+    )
+
+    def partition_oc_by_clear_state(
```

Comment:
Re-using this name confused me since there is a method by this name in `CallConfigs`

Reply:
Given the functionality, maybe the name `clear_n_others` make sense?

Reply:
makes sense